### PR TITLE
[testing] make @dtypes mandatory when using @dtypesIf

### LINKED ISSUE #53647: dtypesIfCUDA is footgun for non-CUDA devices (see test_linspace_special_steps)
(TBF)

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/53647

With this if a test forgets to add `@dtypes` while using `@dtypesIf`, following error is raised
```
AssertionError: @dtypes is mandatory when using @dtypesIf however 'test_exponential_no_zero' didn't specify it
```

**Tested Locally**

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -680,6 +680,7 @@ def test_shape(m, n, p, nnz, broadcast, index_dtype, alpha_beta=None):
                   *((torch.complex128,) if CUSPARSE_SPMM_COMPLEX128_SUPPORTED else ()),
                   *torch.testing.get_all_fp_dtypes(include_bfloat16=SM80OrLater,
                                                    include_half=SM53OrLater))
+    @dtypes(())  # since dtypes is mandatory
```
Comment (kshitij12345) [inline test/test_sparse_csr.py:683]: Surprised to know that you can pass `()` to `dtypes`

Comment (kshitij12345) [reply test/test_sparse_csr.py:683]: It fails for XLA with 
```
Nov 11 17:41:16   File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 529, in instantiate_test
Nov 11 17:41:16     if len(dtype_combination) > 1 else dtype_combination[0])
Nov 11 17:41:16 IndexError: tuple index out of range
```

At the following line : https://github.com/pytorch/xla/blob/2014c477ef593561a8bdb45d8e94cde93456a428/test/pytorch_test_base.py#L529

Maybe just use `@dtypesIfCUDA` for `@dtypes` on tests marked with `onlyCUDA`.

Comment (kshitij12345) [reply test/test_sparse_csr.py]: Using @dtypes() works.

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

File: test/test_nn.py [hunk 2]
```diff
@@ -15331,6 +15331,7 @@ def test_softmax_results(self, device, dtype):
                         self.assertEqual(input.grad, ref_input.grad)
 
     @onlyCUDA
+    @dtypes()  # since dtypes is mandatory
```
Comment (mruberry) [inline test/test_nn.py:15334]: For these cases let's replace `@dtypesIfCUDA` with `@dtypes`

**Review by mruberry — APPROVED** [review]
Cool! Just see one inline comment about updating existing tests that exclusively use `@dtypesIf...`

File: test/test_nn.py (hunk 2 above)
Comment (kshitij12345) [reply test/test_nn.py:15334]: Ofcourse! Should have thought about that! Thank you very much!

**Comment by kshitij12345** [comment]
@mruberry this should be ready now :)

**Pull request closed without merging**